prompting/apparmor/message.go

```diff
@@ -0,0 +1,452 @@
+package apparmor
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+
+	"golang.org/x/xerrors"
+)
+
+// overwrite implements io.Writer that writes over an existing buffer.
+//
+// It is used to perform in-place modifications of a larger memory buffer.
+type overwrite struct {
+	Buffer []byte
+	Offset int
+}
+
+// Write overwrites the buffer at a given offest.
```

Comment:
I also don't understand this: why is `Offset` never modified? And the `overwriter` struct is private and not the return type of any function, so it seems like it would never be changed externally either.  In fact, I don't really see why the `Offset` field is even needed at all, or `overwriter` at all, for that matter.

The only time the `overwriter` struct is actually used is in:

https://github.com/olivercalder/snapd/blob/4eec411b2380b418364502257d1b4c9abfbba8f9/prompting/apparmor/message.go#L69-L74

and the return value is the raw buffer itself, nothing related to the fact that it is part of an `overwriter`, so I'm unsure why it's really necessary. For example, why couldn't a `bytes.Buffer` be created, used for the `binary.Write()` call, and then quickly discarded in the same way the `overwriter` is here. I suppose the benefit is that we can construct the `overwriter` with the preallocated buffer we want, which wouldn't be possible otherwise. But that still begs the question why the `Offset` field is required at all, and not to just always operate on the front of the buffer (`Offset = 0`).

The one thing I can think of is that we always want to keep the header in place, and overwrite after it. In the excerpt above, a header is written to the beginning of the buffer. The `RequestBuffer()` function was only used in one place:

https://github.com/olivercalder/snapd/blob/prompting/prompting/notifier/notifier.go#L199-L210

and here, this is because the ioctl syscall expects some information to be encoded at the memory location passed to the syscall (the address of the start of the buffer). So the buffer is overwritten by the syscall, and we don't need the offset there either.

In summary, I think we don't actually need an `Offset` field in the struct, and I think we should make it more clear that the `overwriter` type is really just a minimal wrapper over a preallocated buffer to allow it to be used as an `io.Writer`. I'll make the changes under these assumptions for now.

Reply:
I looked at this again (sorry, it just bugged me) and I guess this:
```go
func RequestBuffer() []byte {
	buf := bytes.NewBuffer(nil)
	buf.Grow(0xFFFF)
	header := MsgHeader{Version: 2, Length: uint16(0xFFFF)}
	binary.Write(buf, binary.LittleEndian, &header)
	buf.Write(make([]byte, 0xFFFF-buf.Len()))
	return buf.Bytes()
}
```
avoids the new type and does not feel less readable and has the same amount of allocation. Wdyt?

Reply:
Thank you for thinking more on this, I really like your idea. I think technically this will likely need to allocate the initial buffer and then reallocate on the larger write, but that seems completely worth it to me to avoid the new single-use type. It's really only about being able to use `binary.Write()` on a `[]byte` before returning it, so I think a bit of extra gymnastics within this function is preferable to anything extra outside this function. I'll go with this technique.

Reply:
I believe this would solve the problem without reallocation, but it abuses/breaks some assumptions of the `bytes.Buffer` type:
```
func RequestBuffer() []byte {
	backingBuf := make([]byte, 0xFFFF)
	buf := bytes.NewBuffer(backingBuf[:0]) // reset size but not capacity, with backingBuf pre-zeroed
	header := MsgHeader{Version: 2, Length: uint16(0xFFFF)}
	binary.Write(buf, binary.LittleEndian, &header)
	return buf.Bytes()[:0xFFFF] // set capacity back to the full size of backingBuf
}
```
This violates the ownership statement in https://pkg.go.dev/bytes#NewBuffer and assumes that `buf` continues to use the same `backingBuf` without changing the remaining contents, which I believe is true. So technically this should work. But it's a lot messier.

Otherwise, I don't think there's a way to avoid allocating $$O(2*0xFFFF)$$ bytes. Here's another possible, slightly less messy option:
```
func RequestBuffer() []byte {
	backingBuf := make([]byte, 0xFFFF)
	buf := bytes.NewBuffer(backingBuf[:0]) // reset size but not capacity, with backingBuf pre-zeroed
	header := MsgHeader{Version: 2, Length: uint16(0xFFFF)}
	binary.Write(buf, binary.LittleEndian, &header)
	buf.Write(backingBuf[buf.Len():], 0xFFFF-buf.Len()) // set length of buf the "right" way, but without a second allocation
	return buf.Bytes()
}
```

I assume these are both not really up to the standards we want, however. Is an extra `calloc()` and `memcpy()` of $$O(0xFFFF)$$ acceptable? Probably.